daemon/mgr/container_utils.go

```diff
@@ -105,35 +107,45 @@ func (mgr *ContainerManager) generateName(id string) string {
 	return name
 }
 
-// getRuntime returns runtime real path.
-// TODO(huamin): do we need validate runtime is executable ?
-func (mgr *ContainerManager) getRuntime(runtime string) (string, error) {
+// getRuntimeType returns containerd runtime type, type shim v1 by default.
+func (mgr *ContainerManager) getRuntimeType(runtime string) (string, error) {
 	r, exist := mgr.Config.Runtimes[runtime]
 	if !exist {
 		return "", fmt.Errorf("failed to find runtime %s in daemon config", runtime)
 	}
+	if r.Type == "" {
+		return ctrd.RuntimeTypeRuncV1, nil
+	}
+	return r.Type, nil
+}
 
-	// it is ok to use runtime name as a path.
-	rPath := runtime
-	// generally speaking, path is not be empty, but we not forbid empty path
-	// in config set, since name can be a path too.
-	if r.Path != "" {
-		rPath = r.Path
+// generateRuntimeOptions generate options from daemon runtime configurations.
+func (mgr *ContainerManager) generateRuntimeOptions(runtime string) (interface{}, error) {
+	r, exist := mgr.Config.Runtimes[runtime]
+	if !exist {
+		return nil, fmt.Errorf("failed to find runtime %s in daemon config", runtime)
 	}
 
-	// if Runtime has args, use script path as runtime path.
-	if len(r.RuntimeArgs) > 0 {
-		rPath = filepath.Join(mgr.Config.HomeDir, RuntimeDir, runtime)
+	var options interface{}
+	switch o := r.Options.(type) {
+	case *runctypes.RuncOptions:
+		options = &runctypes.RuncOptions{
+			Runtime:       r.Path,
+			RuntimeRoot:   ctrd.RuntimeRoot,
+			CriuPath:      o.CriuPath,
+			SystemdCgroup: mgr.Config.UseSystemd(),
+		}
```

Comment:
generate a brand new options according to the configuration which were passed from config file, thus, there is no data race.
we directly replace some fields here, because these fields are allowed to be configurable directly.